CMakeLists.txt

```diff
@@ -132,20 +108,63 @@ celeritas_setup_option(CELERITAS_CORE_GEO ORANGE)
 celeritas_setup_option(CELERITAS_CORE_GEO Geant4 _allow_g4)
 celeritas_define_options(CELERITAS_CORE_GEO "Celeritas runtime geometry")
 
-if(NOT CELERITAS_DEBUG OR CELERITAS_USE_VecGeom)
-  set(_default_lock ON)
-else()
-  set(_default_lock OFF)
+#----------------------------------------------------------------------------#
+# CMAKE VERSION CHECKS
+#----------------------------------------------------------------------------#
+
+if(CMAKE_VERSION VERSION_LESS 3.13 AND CELERITAS_USE_CUDA AND CELERITAS_USE_MPI)
+  message(FATAL_ERROR "Celeritas requires CMake 3.13 or higher "
+    "when building with CUDA + MPI.")
+endif()
+if(CMAKE_VERSION VERSION_LESS 3.18 AND CELERITAS_USE_CUDA
+    AND CELERITAS_USE_VecGeom)
+  message(FATAL_ERROR "VecGeom+CUDA requires CMake 3.18 or higher to support "
+    "\"Separable compilation\".")
+endif()
+if(CMAKE_VERSION VERSION_LESS 3.18 AND CMAKE_CUDA_ARCHITECTURES)
+  message(FATAL_ERROR "The CMAKE_CUDA_ARCHITECTURES flag is not compatible "
+    "with this version of CMake. Set CMAKE_CUDA_FLAGS.")
+endif()
+if(CMAKE_VERSION VERSION_LESS 3.22 AND CELERITAS_USE_HIP)
+  message(WARNING "HIP support is immature; CMake 3.22+ is recommended.")
+endif()
+
+#----------------------------------------------------------------------------#
+# CMAKE INTRINSIC OPTIONS
+#
+# These are generally used to initialize properties on targets, and it's
+# possible Celeritas is being built inside another project. Instead of saving
+# these as cache variables (which change the defaults project-wide, including
+# changing behavior of other code that was loaded before Celeritas) set them as
+# local variables to be inherited underneath Celeritas.
+#----------------------------------------------------------------------------#
+
+### Build flags ###
+if(DEFINED CMAKE_BUILD_TYPE AND NOT CMAKE_BUILD_TYPE)
+  if(CELERITAS_DEBUG)
+    set(_default_build_type "Debug")
+  else()
+    set(_default_build_type "Release")
+  endif()
+  set(CMAKE_BUILD_TYPE "${_default_build_type}" CACHE STRING "Build type" FORCE)
+  message(STATUS "Set default CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}")
+  unset(_default_build_type)
+endif()
+celeritas_set_default(CMAKE_CXX_STANDARD 17)
+celeritas_set_default(CMAKE_CXX_EXTENSIONS OFF)
+if(CELERITAS_USE_CUDA)
+  celeritas_set_default(CMAKE_CUDA_STANDARD ${CMAKE_CXX_STANDARD})
+  celeritas_set_default(CMAKE_CUDA_EXTENSIONS ${CMAKE_CXX_EXTENSIONS})
 endif()
-cmake_dependent_option(CELERITAS_TEST_RESOURCE_LOCK
-  "Only run one GPU-enabled test at a time" "${_default_lock}"
-  "CELERITAS_BUILD_TESTS" OFF
-)
 
-if(NOT DEFINED CMAKE_CXX_STANDARD)
-  # Default CMake standard to C++17 for this directory and underneath
-  set(CMAKE_CXX_STANDARD 17)
-  set(CMAKE_CXX_EXTENSIONS OFF)
+### Linking flags ###
+# Default to building shared libraries
+celeritas_set_default(BUILD_SHARED_LIBS ON)
```

Comment:
Yes, I have mixed feelings about it. It's pretty common for projects to define this as an "option", even though it simply changes the default for a cmake property. From the cmake docs:
> This variable is often added to projects as an option so that each user of a project can decide if they want to build the project using shared or static libraries.

My personal view is that it's confusing to define these cmake internal variables as cache variables: there's weird inconsistencies between the cmake variable scoping, the "defaulting" nature of these commands, and the fact that CMake defines the commands internally but `option` and `set(... CACHE` forces us to provide our own documentation.

Defining one of these as cache variables also makes *us* to choose the line between "important enough for the user to see in the GUI" and not. Do we want to also define GUI options for `CMAKE_EXPORT_COMPILE_COMMANDS`, for example.

Reply:
On the other hand, I did not see it as deciding what is "important enough for the user to see in the GUI" but rather show in the GUI what is either customizable (`CELERITAS_`) OR we changed the default compared to `'bare' CMake`.  I.e. an easy way to document all that is "not the default behavior"/"out of the ordinary".

Reply:
True. We could define a `CELERITAS_BUILD_SHARED_LIBS` that defaults to `BUILD_SHARED_LIBS` or TRUE, and then do 
```cmake
set(BUILD_SHARED_LIBS ${CELERITAS_BUILD_SHARED_LIBS})
```
inside our top cmake... but this feels like overcomplicating things.

I think "changing the default" is OK but don't feel too strongly here.

Reply:
I agree that just 'changing the default' is fine and that the `CELERITAS_BUILD_SHARED_LIBS` might be too much.  However I would add to the final commit a note that make it explicit that the change was intentional or at least a known side effect.